### PR TITLE
fixed a potential bug in CNN ROI RawWaveformDump

### PR DESCRIPTION
**Pull request by wwuoneway**
fixed a potential bug by adding a protection to avoid the same sim channel in an event to be saved more than once

### CODE REVIEW

**Comment by lgarren** [comment]
trigger build

**Comment by chalt007** [comment]
All CI builds are OK.

**Comment by chalt007** [comment]
approve

File: larrecodnn/ImagePatternAlgs/Tensorflow/Modules/RawWaveformDump_module.cc [hunk 1]
```diff
@@ -464,6 +464,9 @@ nnet::RawWaveformDump::analyze(art::Event const& evt)
 
     } // loop over SimChannels
 
+    std::vector<raw::ChannelID_t> SelectedChannel;
+    SelectedChannel.clear();
```
Comment (knoepfel) [inline larrecodnn/ImagePatternAlgs/Tensorflow/Modules/RawWaveformDump_module.cc:468]: Searching through an unsorted vector using `std::find` can be inefficient.  In this case, I would recommend using an `std::set`:
```.cpp
std::set<raw::ChannelID_t> selected_channels;
```
Then see below.  Note that it is not necessary to call `clear` on a newly-created vector.

File: larrecodnn/ImagePatternAlgs/Tensorflow/Modules/RawWaveformDump_module.cc [hunk 2]
```diff
@@ -472,6 +475,13 @@ nnet::RawWaveformDump::analyze(art::Event const& evt)
           rndm_dist(rndm_engine); // randomly select one channel with a signal from this particle
         chnum = ittrk.second[i];
 
+        if (std::find(SelectedChannel.begin(), SelectedChannel.end(), chnum) != SelectedChannel.end()) {
+          continue; // channel already used
+        }
+        else {
+          SelectedChannel.push_back(chnum);
+        }
+
```
Comment (knoepfel) [inline larrecodnn/ImagePatternAlgs/Tensorflow/Modules/RawWaveformDump_module.cc:484]: Using the `selected_channels` object, the above lines can be replaced with:
```.cpp
if (not selected_channels.insert(chnum).second) {
  continue;
}
```
The above code will attempt to insert the `chnum` value into `selected_channels`.  If the insertion succeeds, then the value of `.second` will be `true`.  If the insertion fails, then `.second` will return `false`.

Comment (wwuoneway) [reply larrecodnn/ImagePatternAlgs/Tensorflow/Modules/RawWaveformDump_module.cc]: Hi Kyle, thank you very much for your comments. I already replaced the codes as you recommended. I did learn a lot from your suggestions.

**Comment by knoepfel** [comment]
trigger build

**Comment by knoepfel** [comment]
No problem, @wwuoneway.  Assuming the CI tests come back positive, I will approve the PR.

**Comment by lgarren** [comment]
The CI tests are normal

**Comment by lgarren** [comment]
approve

**Pull request merged**